src/bthread/bthreadxx.h

```diff
@@ -0,0 +1,284 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Authors: Shuo Zang (jasonszang@126.com)
+
+#ifndef BTHREAD_BTHREADXX_H
+#define BTHREAD_BTHREADXX_H
+
+#include <butil/macros.h>
+
+#ifdef BUTIL_CXX11_ENABLED
+
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <system_error>
+#include <type_traits>
+#include "bthread/bthread.h"
+#include "bthread/condition_variable.h"
+#include "bthread/mtx_cv_base.h"
+
+namespace bthread {
+
+namespace detail {
+
+template<typename MatchType, typename TArg>
+using DisableOverload = std::enable_if<
+        !std::is_base_of<
+                MatchType,
+                typename std::decay<TArg>::type
+        >::value
+>;
+
+template<typename MatchType, typename TArg>
+using DisableOverloadT = typename DisableOverload<MatchType, TArg>::type;
+
+// Just for identifying bthread. There is a bthread_id_t but it is a totally different thing.
+using bthread_id = bthread_t;
+
+constexpr bthread_t NULL_BTHREAD = 0;
+
+struct ThreadFunc {
+    virtual ~ThreadFunc() = default;
+
+    virtual void run() = 0;
+};
+
+template<typename Function>
+struct ThreadFuncImpl : public ThreadFunc {
+    explicit ThreadFuncImpl(Function&& f) : f_(std::forward<Function>(f)) {
+    }
+
+    void run() override {
+        f_();
+    }
+
+    Function f_;
+};
+
+template<typename Callable>
+std::unique_ptr<ThreadFunc> make_func_ptr(Callable&& f) {
+    return std::unique_ptr<ThreadFunc>(new ThreadFuncImpl<Callable>(std::forward<Callable>(f)));
+}
+
+inline void* thread_func_proxy(void* owning_func_ptr) {
+    std::unique_ptr<ThreadFunc> func_ptr{static_cast<ThreadFunc*>(owning_func_ptr)};
+    func_ptr->run();
+    return nullptr;
+}
+
+} // namespace detail
+
+class BThreadIdWrapper;
+
+namespace this_bthread {
```

Comment:
Consider renaming this_bthread to this_thread, with "bthread::" as the namespace, the full name should be clear to users.